File: typescript/linked-list/linked-list.ts
class Node<T> {
  value: T
  next: Node<T> | undefined
  prev: Node<T> | undefined
  constructor(value: T) { this.value = value }
}

class LinkedList<T> {
  head: Node<T> | undefined
  tail: Node<T> | undefined

  count(): number {
    let count = 0
    let node = this.head
    while (node !== undefined) {
      node = node.next
      count += 1
    }
    return count
  }

  push(value: T): LinkedList<T> {
    const node = new Node<T>(value)
    if (this.tail === undefined) {
      this.head = node
    } else {
      this.tail.next = node
      node.prev = this.tail
    }
    this.tail = node
    return this
  }

  unshift(value: T): LinkedList<T> {
    const node = new Node<T>(value)
    if (this.head === undefined) {
      this.tail = node
    } else {
      this.head.prev = node
      node.next = this.head
    }
    this.head = node
    return this
  }

  pop(): T {
    const node = this.tail
    if (node === undefined) {
      throw new Error('List is empty')
    }
    if (node.prev === undefined) {
      this.tail = undefined
      this.head = undefined
    } else {
      node.prev.next = undefined
      this.tail = node.prev
    }
    return node.value
  }

  shift(): T {
    const node = this.head
    if (node === undefined) {
      throw new Error('List is empty')
    }
    if (node.next === undefined) {
      this.tail = undefined
      this.head = undefined
    } else {
      node.next.prev = undefined
      this.head = node.next
    }
    return node.value
  }

  delete(value: T): LinkedList<T> {
    let node = this.head
    while (node !== undefined) {
      if (node.value !== value) {
        node = node.next
        continue
      }

      if (node.next) {
        node.next.prev = node.prev
      } else {
        this.tail = node.prev
      }

      if (node.prev) {
        node.prev.next = node.next
      } else {
        this.head = node.next
      }

      break
    }
    return this
  }
}

export default LinkedList

File: lua/linked-list/linked-list.lua
local LinkedList = {}
LinkedList.__index = LinkedList
setmetatable(LinkedList, {
    __call = function(cls, ...) return cls:new(...) end,
})

function LinkedList:new(...)
    return setmetatable({}, self)
end

function LinkedList:push(value)
    local node = {value=value}
    if self.tail then
        node.prev = self.tail
        self.tail.next = node
    else
        self.head = node
    end
    self.tail = node
    return self
end

function LinkedList:pop()
    local value
    local node = self.tail
    if node then
        value = node.value
        if node.prev then
            node.prev.next = nil
            self.tail = node.prev
        else
            self.tail = nil
            self.head = nil
        end
    end
    return value
end

function LinkedList:unshift(value)
    local node = {value=value}
    if self.head then
        node.next = self.head
        self.head.prev = node
    else
        self.tail = node
    end
    self.head = node
    return self
end

function LinkedList:shift()
    local value
    local node = self.head
    if node then 
        value = node.value
        if node.next then
            node.next.prev = nil
            self.head = node.next
        else
            self.tail = nil
            self.head = nil
        end
    end
    return value
end

function LinkedList:count()
    local count = 0
    local node = self.head
    while node do
        count = count + 1
        node = node.next
    end
    return count
end

-- delete *all* nodes with matching value
function LinkedList:delete(value)
    local node = self.head
    while node do
        if node.value == value then
            -- unlink this node
            if node.next then
                node.next.prev = node.prev
            else
                self.tail = node.prev
            end
            if node.prev then
                node.prev.next = node.next
            else
                self.head = node.next
            end
        end
        node = node.next
    end
end

return LinkedList
